manage-github.py

```diff
@@ -158,74 +118,95 @@ def ensure_teams(config, org):
 
-    print('Updating team repositories...')
+    print('Checking org repositories...')
     for repo in opensafely.repos.values():
         # admins have access to all repos
-        managers.add_repo(repo, 'admin')
+        yield from managers.add_repo(repo, 'admin')
 
         # either a protected developer repo or not
         if repo.full_name in config['protected_repositories']:
             # stricter branch protection for these repos 
-            protect_branch(
+            yield from protect_branch(
                 repo,
                 enforce_admins=True,
                 required_approving_review_count=1,
             )
-            developers.add_repo(repo, 'admin')
-            researchers.add_repo(repo, 'triage')
+            yield from developers.add_repo(repo, 'admin')
+            yield from researchers.add_repo(repo, 'triage')
         else:
             # basic branch protection against force pushes, even for admins
             protect_branch(repo, enforce_admins=True)
             researchers.add_repo(repo, 'admin')
 
 
-if __name__ == '__main__':
-    args = parser.parse_args()
-    EXECUTE = args.execute
-
-    config = yaml.safe_load(open(args.config))
+def main(argv=sys.argv[1:]):
+    parser = argparse.ArgumentParser(
+        description='Apply policy to OpenSAFELY github org'
+    )
+    parser.add_argument('config', help='The team config')
+    parser.add_argument('--exec', action='store_true',
+                        dest='execute',
+                        help='Automatically execute commands')
+    parser.add_argument('--dry-run', action='store_true',
+                        dest='dry_run',
+                        help='Just print what would change and exit')
 
-    if 'GITHUB_TOKEN' not in os.environ:
-        print('Error: missing environment variable GITHUB_TOKEN')
-        print('You need a Personal Access Token, with the admin:org and all repo permssions')
-        print('https://docs.github.com/en/github/authenticating-to-github/creating-a-personal-access-token')
-        sys.exit(1)
+    args = parser.parse_args(argv)
 
-    gh = Github(os.environ['GITHUB_TOKEN'])
-    org = gh.get_organization('OpenSAFELY')
+    org = client.get_org()
+    config = yaml.safe_load(open(args.config))
 
-    if not EXECUTE:
+    if args.dry_run:
         print('*** DRY RUN - no changes will be made ***')
 
-    ensure_teams(config, org)
+    changes = []
+    for change in ensure_teams(config, org, args.execute):
+        print(change)
+        if args.execute:
+            change()
+        elif not self.dry-run:
+            changes.append(change)
+
+    if changes:
+        print("Do you want to apply the above changes (y/n)?")
+        i, _, _ = select.select([sys.stdin], [], [], 5)
+        if i:
+            if sys.stdin.readline().strip().lower() == 'y':
+                for change in changes:
+                    print(change)
+                    change()
+    else:
+        print('No changes needed')
```

Comment:
not quite sure about this block - if `args.execute` is true, `changes` will be `[]` (falsey), so the path through the code will do
```
print(change)
change()
print('No changes needed')
```

Reply:
moving to updated commit